rust-mozjs/examples/wasm.rs

```diff
@@ -0,0 +1,180 @@
+    rooted!(in(rt.cx()) let mut wasm = UndefinedValue());
+    rooted!(in(rt.cx()) let mut wasm_module = UndefinedValue());
+    rooted!(in(rt.cx()) let mut wasm_instance = UndefinedValue());
+
+    unsafe {
+        assert!(JS_GetProperty(
+            rt.cx(),
+            global.handle(),
+            b"WebAssembly\0".as_ptr() as *const c_char,
+            &mut wasm.handle_mut()
+        ));
+        rooted!(in(rt.cx()) let mut wasm_obj = wasm.to_object());
+        assert!(JS_GetProperty(
+            rt.cx(),
+            wasm_obj.handle(),
+            b"Module\0".as_ptr() as *const c_char,
+            &mut wasm_module.handle_mut()
+        ));
+        assert!(JS_GetProperty(
+            rt.cx(),
+            wasm_obj.handle(),
+            b"Instance\0".as_ptr() as *const c_char,
+            &mut wasm_instance.handle_mut()
+        ));
+
+        // Construct Wasm module from bytes.
+        rooted!(in(rt.cx()) let mut module = null_mut::<JSObject>());
+        {
+            let array_buffer = JS::NewArrayBufferWithUserOwnedContents(
+                rt.cx(),
+                HI_WASM.len(),
+                HI_WASM.as_ptr() as _,
+            );
+            assert!(!array_buffer.is_null());
+
+            rooted!(in(rt.cx()) let val = ObjectValue(array_buffer));
+            let args = HandleValueArray {
+                length_: 1,
+                elements_: &*val,
+            };
+
+            assert!(Construct1(
+                rt.cx(),
+                wasm_module.handle(),
+                &args,
+                &mut module.handle_mut()
+            ))
+        }
+
+        // Construct Wasm module instance with required imports.
+        rooted!(in(rt.cx()) let mut instance = null_mut::<JSObject>());
+        {
+            // Build "env" imports object.
+            rooted!(in(rt.cx()) let mut env_import_obj = JS_NewPlainObject(rt.cx()));
+            assert!(!env_import_obj.is_null());
+            let function = JS_DefineFunction(
+                rt.cx(),
+                env_import_obj.handle().into(),
+                b"bar\0".as_ptr() as *const c_char,
+                Some(bar),
+                1,
+                0,
+            );
+            assert!(!function.is_null());
+            rooted!(in(rt.cx()) let mut env_import = ObjectValue(env_import_obj.get()));
+            // Build imports bag.
+            rooted!(in(rt.cx()) let mut imports = JS_NewPlainObject(rt.cx()));
+            assert!(!imports.is_null());
+            assert!(JS_SetProperty(
+                rt.cx(),
+                imports.handle(),
+                b"env\0".as_ptr() as *const c_char,
+                env_import.handle()
+            ));
+
+            rooted!(in(rt.cx()) let mut arg1 = ObjectValue(module.get()));
+            rooted!(in(rt.cx()) let mut arg2 = ObjectValue(imports.get()));
+            let args = HandleValueArray::from_rooted_slice(&[arg1.get(), arg2.get()]);
```

Comment:
Let's use rooted_vec here; it will be safe to call `HandleValueArray::from_rooted_slice` using the slice from inside the vector.

Reply:
To put this in rooted_vec we would need to wrap JSval in Heap (beacuse it needs to implement Traceable).

Reply:
maybe we should port RootedValueArray